File: clrhost/src/altv-c-api/checkpoint.cpp
#include "checkpoint.h"

// Entity

uint16_t Checkpoint_GetID(alt::ICheckpoint* checkpoint) {
    return checkpoint->GetID();
}

alt::Position Checkpoint_GetPosition(alt::ICheckpoint* checkpoint) {
    return checkpoint->GetPosition();
}

void Checkpoint_SetPosition(alt::ICheckpoint* checkpoint, alt::Position pos) {
    checkpoint->SetPosition(pos);
}

alt::Rotation Checkpoint_GetRotation(alt::ICheckpoint* checkpoint) {
    return checkpoint->GetRotation();
}

void Checkpoint_SetRotation(alt::ICheckpoint* checkpoint, alt::Rotation rot) {
    checkpoint->SetRotation(rot);
}

uint16_t Checkpoint_GetDimension(alt::ICheckpoint* checkpoint) {
    return checkpoint->GetDimension();
}

void Checkpoint_SetDimension(alt::ICheckpoint* checkpoint, uint16_t dimension) {
    checkpoint->SetDimension(dimension);
}

void Checkpoint_GetMetaData(alt::ICheckpoint* checkpoint, const char* key, alt::MValue &val) {
    val = checkpoint->GetMetaData(key);
}

void Checkpoint_SetMetaData(alt::ICheckpoint* checkpoint, const char* key, alt::MValue* val) {
    checkpoint->SetMetaData(key, *val);
}

void Checkpoint_GetSyncedMetaData(alt::ICheckpoint* checkpoint, const char* key, alt::MValue &val) {
    val = checkpoint->GetSyncedMetaData(key);
}

void Checkpoint_SetSyncedMetaData(alt::ICheckpoint* checkpoint, const char* key, alt::MValue* val) {
    checkpoint->SetSyncedMetaData(key, *val);
}

// Checkpoint

bool Checkpoint_IsGlobal(alt::ICheckpoint* checkpoint) {
    return checkpoint->IsGlobal();
}

uint8_t Checkpoint_GetCheckpointType(alt::ICheckpoint* checkpoint) {
    return checkpoint->GetCheckpointType();
}

float Checkpoint_GetHeight(alt::ICheckpoint* checkpoint) {
    return checkpoint->GetHeight();
}

float Checkpoint_GetRadius(alt::ICheckpoint* checkpoint) {
    return checkpoint->GetRadius();
}

alt::RGBA Checkpoint_GetColor(alt::ICheckpoint* checkpoint) {
    return checkpoint->GetColor();
}

File: src/AltV.Net/Elements/Entities/Checkpoint.cs
using System;
using AltV.Net.Data;
using AltV.Net.Elements.Args;
using AltV.Net.Native;

namespace AltV.Net.Elements.Entities
{
    public class Checkpoint : WorldObject, ICheckpoint
    {
        public override Position Position
        {
            get
            {
                CheckExistence();
                var position = Position.Zero;
                AltVNative.Checkpoint.Checkpoint_GetPosition(NativePointer, ref position);
                return position;
            }
            set
            {
                CheckExistence();
                AltVNative.Checkpoint.Checkpoint_SetPosition(NativePointer, value);
            }
        }

        public override short Dimension
        {
            get
            {
                CheckExistence();
                return AltVNative.Checkpoint.Checkpoint_GetDimension(NativePointer);
            }
            set
            {
                CheckExistence();
                AltVNative.Checkpoint.Checkpoint_SetDimension(NativePointer, value);
            }
        }

        public IPlayer Target
        {
            get
            {
                CheckExistence();
                var playerPointer = AltVNative.Checkpoint.Checkpoint_GetTarget(NativePointer);
                if (playerPointer == IntPtr.Zero) return null;
                return Alt.Module.PlayerPool.GetOrCreate(playerPointer, out var player) ? player : null;
            }
        }

        public override void SetMetaData(string key, object value)
        {
            CheckExistence();
            var mValue = MValue.CreateFromObject(value);
            AltVNative.Checkpoint.Checkpoint_SetMetaData(NativePointer, key, ref mValue);
        }

        public override bool GetMetaData<T>(string key, out T result)
        {
            CheckExistence();
            var mValue = MValue.Nil;
            AltVNative.Checkpoint.Checkpoint_GetMetaData(NativePointer, key, ref mValue);
            if (!(mValue.ToObject() is T cast))
            {
                result = default;
                return false;
            }

            result = cast;
            return true;
        }

        public bool IsGlobal
        {
            get
            {
                CheckExistence();
                return AltVNative.Checkpoint.Checkpoint_IsGlobal(NativePointer);
            }
        }

        public byte CheckpointType
        {
            get
            {
                CheckExistence();
                return AltVNative.Checkpoint.Checkpoint_GetType(NativePointer);
            }
        }

        public float Height
        {
            get
            {
                CheckExistence();
                return AltVNative.Checkpoint.Checkpoint_GetHeight(NativePointer);
            }
        }

        public float Radius
        {
            get
            {
                CheckExistence();
                return AltVNative.Checkpoint.Checkpoint_GetRadius(NativePointer);
            }
        }

        public Rgba Color
        {
            get
            {
                CheckExistence();
                var color = Rgba.Zero;
                AltVNative.Checkpoint.Checkpoint_GetColor(NativePointer, ref color);
                return color;
            }
        }

        public Checkpoint(IntPtr nativePointer) : base(nativePointer, BaseObjectType.Checkpoint)
        {
        }

        public bool IsPlayerIn(IPlayer player)
        {
            CheckExistence();
            if (!player.Exists)
            {
                throw new EntityDeletedException(player);
            }

            return AltVNative.Checkpoint.Checkpoint_IsPlayerIn(NativePointer, player.NativePointer);
        }

        public bool IsVehicleIn(IVehicle vehicle)
        {
            CheckExistence();
            if (!vehicle.Exists)
            {
                throw new EntityDeletedException(vehicle);
            }

            return AltVNative.Checkpoint.Checkpoint_IsVehicleIn(NativePointer, vehicle.NativePointer);
        }
        
        public void Remove()
        {
            Alt.RemoveCheckpoint(this);
        }
    }
}